spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -45,36 +45,61 @@
  */
 public class PubSubTemplate implements PubSubOperations, InitializingBean {
 
+	private static final String DEFAULT_SOURCE_NAME = "spring";
+
 	private final String projectId;
 
 	private final GoogleCredentials credentials;
 
 	private ConcurrentHashMap<String, Publisher> publishers = new ConcurrentHashMap<>();
 
+	/**
+	 * Aimed at reusing the same executor across every {@link Publisher}, so the number of allocated
+	 * threads doesn't blow out of proportion.
+	 */
 	private ExecutorProvider executorProvider;
 
 	private MessageConverter messageConverter = new SimpleMessageConverter();
 
+	/**
+	 * The Spring component a Google Cloud Pub/Sub call is originated from.
+	 *
+	 * <p>Spring Integration callers should set this to "spring-integration", Spring Cloud to
+	 * "spring-cloud", etc.
+	 */
+	private String sourceName = DEFAULT_SOURCE_NAME;
+
+	/**
+	 * The version of the Spring component a Google Cloud Pub/Sub call is originated from.
+	 *
+	 * <p>By default, the version is obtained from this package's manifest file, which is mirrored
+	 * from this package's pom.xml file. Ideally, this would be the version of the library invoking
+	 * this call (e.g., Spring Integration, Spring Cloud, etc.).
+	 */
+	private String sourceVersion;
+
 	private int concurrentProducers = 1;
 
 	public PubSubTemplate(GoogleCredentials credentials, String projectId) {
 		this.projectId = projectId;
 		this.credentials = credentials;
+		this.sourceVersion = this.getClass().getPackage().getImplementationVersion();
 		this.executorProvider = InstantiatingExecutorProvider.newBuilder()
 				.setExecutorThreadCount(this.concurrentProducers).build();
 	}
 
 	@Override
 	public ListenableFuture<String> send(final String topic, Message message) {
-
-		Publisher publisher = this.publishers.computeIfAbsent(topic, s -> {
+		// Reuse existing publisher if one already exists.
+		Publisher publisher = this.publishers.computeIfAbsent(topic, key -> {
 			try {
-				return Publisher.defaultBuilder(TopicName.create(this.projectId, topic))
+				return Publisher.defaultBuilder(TopicName.create(this.projectId, key))
 						.setExecutorProvider(this.executorProvider)
```

Comment:
H-m. I didn't think so far, but if that is the way you, guys, see it, let's go that way then!
We can merge as is right now and address `PublisherFactory` in the separate issue.

Yes, that was the reason why we chose `Factory` abstraction for Apache Kafka `Producer` and `Consumer`. Jut because they are volatile and non-inversion of control entities.